loki/transform/transform_inline.py

```diff
@@ -238,17 +248,38 @@ def _map_unbound_dims(var, val):
                 if isinstance(arg, sym.Array):
                     # Resolve implicit dimension ranges of the passed value,
                     # eg. when passing a two-dimensional array `a` as `call(arg=a)`
-                    qualified_value = val if val.dimensions else val.clone(
-                        dimensions=tuple(sym.Range((None, None)) for _ in arg.shape)
-                    )
+                    # Check if val is a DeferredTypeSymbol, as it does not have a `dimensions` attribute
+                    if not isinstance(val, sym.DeferredTypeSymbol) and val.dimensions:
+                        qualified_value = val
+                    else:
+                        qualified_value = val.clone(
+                            dimensions=tuple(sym.Range((None, None)) for _ in arg.shape)
+                        )
                     arg_vars = tuple(v for v in member_vars if v.name == arg.name)
-                    argmap.update((v, _map_unbound_dims(v, qualified_value)) for v in arg_vars)
+                    argmap.update((v, val.clone(dimensions = _map_unbound_dims(v, qualified_value))) for v in arg_vars)
```

Comment:
I was really trying to avoid using deep-clones in this tight level of loop (it's expensive at the wrong granularity, and we already have a few cases where we're being careless with this :sweat_smile:). Is there a particular reasons this is needed here?

Reply:
Ok nevermind, I managed to make that work through ArraySubscripts.